### PR TITLE
Extract authentication actions from `UserController`

### PR DESCRIPTION
**Pull request by dgdavid**
Hi there!

TL;DR

This PR is an attempt to extract login/logout actions from `UserController` and put them in its own `SessionController`.

I really hope it could be useful :)

---

There are a couple of things that I would like to clarify why I did it or what I think about it. So, let me explain

- `Naming` In spite of the preferable Rails convention is to use plural for controllers, I saw that you are using both, singular and plural. So, I chose the singular form for the new `SessionController`, just like that one that is being modified, `UserController`. However, if it was up to me, probably I would rename it to `SessionsController`.

- `Class vs Module` I leave `UserSession` class in `app/model` directory (even when it is not exactly a model) after thinking about the possibility to become it in a module and move it to `lib`. But, sincerely,  I am not pretty sure about the convenience of it. 

  In addition, with [current configuration](https://github.com/openSUSE/open-build-service/blob/1cb11f65d9a9fa0fa9bfc9d3a6eefc5da46bc06c/src/api/config/application.rb#L18) and the [default `false` value for `config.eager_load` in the test environment](https://github.com/rails/rails/issues/28736)<sup><a href="#issue-reference-foot-note"> 1</a></sup>, tests fails unless the file was required explicitly (which makes sense).

  So, as I said, `UserSession` was added as a class along with the models, waiting to receive your feedback/thoughts/recommendations about what to do

  - Keep as it
  - Move to `lib` as a module and configure the test environment properly
  - Create the `app/lib` directory for this kind of modules and move there.
  - Other

- `session` As you can see, I leave the `session[:login]` assignation in the controller instead to do it in the `UserSession.create` method because

  - [it seems to be a concern/responsibility of controller](http://guides.rubyonrails.org/action_controller_overview.html#session).
  - I think that passing `session` as an argument of such method could be a code smell.
  - `reset_session` it is also performed by the controller.

- `Scope` I tried to not add changes out of scope the PR. This is the reason why the massive replacement of `User.current[ = nil|.*]` to `UserSession.[current|create|destroy]`, before which is [necessary to change when `DisabledAccount` error is raised](https://github.com/dgdavid/open-build-service/commit/0c1a1ca9741c3ce3eb709c128aa293fc97ee3b94) to keep tests green, [is made in another commit, ready to use](https://github.com/dgdavid/open-build-service/commit/81d72ccea947f4e31e42914a17f77dcc5d69cc9d) if finally changes are approved.

---

<sub><sup id="issue-reference-foot-note">1</sup> Through references to linked issue I realize that you have [an open PR](https://github.com/openSUSE/open-build-service/pull/4727) to set `config.eager_load = true` in test environment.</sub>

### CODE REVIEW

File: src/api/spec/controllers/webui/session_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
+require 'rails_helper'
+
+RSpec.describe Webui::SessionController do
+  let!(:user) { create(:confirmed_user, login: 'tom') }
```
Comment (Ana06) [inline src/api/spec/controllers/webui/session_controller_spec.rb]: use the lazy `let` instead, as this user is not used in `POST #create with deprecated password`.

File: src/api/app/controllers/webui/session_controller.rb [hunk 2]
```diff
@@ -8,36 +8,46 @@ def new; end
   def create
     user = User.find_with_credentials(params[:username], params[:password])
 
-    if user && !user.is_active?
-      redirect_to(root_path, error: 'Your account is disabled. Please contact the administrator for details.')
+    if user.nil?
+      redirect_to(session_new_path, error: 'Authentication failed')
+
```
Comment (Ana06) [inline src/api/app/controllers/webui/session_controller.rb]: why this blank line? blank lines are supposed to be used to separate different conceptual parts of the cope, and this return is related. Please remove this line. :wink:

File: src/api/app/controllers/webui/session_controller.rb [hunk 3]
```diff
@@ -8,36 +8,46 @@ def new; end
   def create
     user = User.find_with_credentials(params[:username], params[:password])
 
-    if user && !user.is_active?
-      redirect_to(root_path, error: 'Your account is disabled. Please contact the administrator for details.')
+    if user.nil?
```
Comment (Ana06) [inline src/api/app/controllers/webui/session_controller.rb]: :see_no_evil: Use `unless user` instead please :pray:

File: src/api/config/routes.rb [hunk 4]
```diff
@@ -328,10 +328,6 @@ def self.public_or_about_path?(request)
       post 'user/register' => :register
       get 'user/register_user' => :register_user
 
-      get 'user/login' => :login
```
Comment (ChrisBr) [inline src/api/config/routes.rb]: Please leave the ``get`` paths for backward compatibility

File: src/api/app/models/user_session.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+class UserSession
+  class FailedAuthentication < StandardError; end
+  class DisabledAccount < StandardError; end
+
+  # Returns the user storaged in current Thread
```
Comment (Ana06) [inline src/api/app/models/user_session.rb]: I think this is evident. :rofl: I am personally against adding comments if they are not strictly needed. If the code is clear, most of the times the comments are not needed. :wink:

File: src/api/app/models/user_session.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  #
+  # @return [User|nil] user for current session or nil if none
+  def self.current
+    Thread.current[:user]
+  end
+
+  # Storages the user in current Thread
+  #
+  # @param [User|nil]
```
Comment (Ana06) [inline src/api/app/models/user_session.rb]: well, I can also give whatever I want to this method. I don't think if it has sense to specify what you expect. Also the variable speaks by itself, so again it can be removed it.

File: src/api/app/models/user_session.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  #
+  # @raise [FailedAuthentication] if not user given
```
Comment (Ana06) [inline src/api/app/models/user_session.rb]: Again, this is evident! There is no different between reading the comments and the code, then the comments can be removed :wink:

File: src/api/spec/models/user_spec.rb [hunk 8]
```diff
@@ -21,6 +21,36 @@
     it { expect(create(:user)).to validate_uniqueness_of(:login).with_message('is the name of an already existing user') }
   end
 
+  describe '#is_active?' do
+    it 'returns true if user is confirmed' do
+      confirmed_user = create(:confirmed_user, login: 'user')
```
Comment (Ana06) [inline src/api/spec/models/user_spec.rb]: why is that not in a `let`? :see_no_evil:

File: src/api/spec/models/user_spec.rb [hunk 9]
```diff
@@ -21,6 +21,36 @@
...
+
+      expect(confirmed_user.is_active?).to be_truthy
+    end
+
+    it 'returns false if user is NOT confirmed' do
+      unconfirmed_user = create(:user, login: 'user')
```
Comment (Ana06) [inline src/api/spec/models/user_spec.rb]: `let` :see_no_evil:

File: src/api/spec/models/user_spec.rb [hunk 10]
```diff
@@ -21,6 +21,36 @@
...
+
+      expect(unconfirmed_user.is_active?).to be_falsey
+    end
+
+    context 'when user has owner' do
+      it 'returns true if owner is confirmed' do
+        confirmed_owner = create(:confirmed_user, login: 'owner')
```
Comment (Ana06) [inline src/api/spec/models/user_spec.rb]: `let` :see_no_evil: and the next line too :pray:

File: src/api/spec/models/user_spec.rb [hunk 11]
```diff
@@ -21,6 +21,36 @@
...
+        user = create(:user, owner: confirmed_owner)
+
+        expect(user.is_active?).to be_truthy
+      end
+
+      it 'returns false if owner not confirmed' do
+        unconfirmed_owner = create(:user, login: 'owner')
```
Comment (Ana06) [inline src/api/spec/models/user_spec.rb]: `let` :see_no_evil: and the next line too :pray:

File: src/api/config/routes.rb (hunk 4 above)
Comment (Ana06) [inline src/api/config/routes.rb]: can we modify the api? or should we keep them at least until the next big release? @adrianschroeter ^^

Comment (ChrisBr) [reply src/api/config/routes.rb]: We need to keep it!

Comment (Ana06) [reply src/api/config/routes.rb]: ok, then I guess we need something like a FIXME to remove them in the next big release. Because we don't want to keep them for ever, right?

File: src/api/app/models/user_session.rb (hunk 5 above)
Comment (dgdavid) [reply src/api/app/models/user_session.rb]: :sweat_smile: :sweat_smile: :sweat_smile: 

You're right.

I did that with the generation of docs in mind since it is something that I am doing for some time now.

That said, I am going to remove them :)

File: src/api/spec/models/user_spec.rb (hunk 8 above)
Comment (dgdavid) [reply src/api/spec/models/user_spec.rb]: Actually, I used a handful of  `let`s  in the `describe` of this test,  but I changed them before making the commit because it seemed to me more explicit and less "noisy" in that way.

Now, after review it to accomplish your requested changed, I think that I found a better way, putting extra needed lets at the top, in the main `describe`.

Indeed, looks better.

File: src/api/config/routes.rb [hunk 12]
```diff
@@ -30,6 +30,7 @@ def self.public_or_about_path?(request)
     binary:       %r{[^\/]*},
     filename:     %r{[^\/]*},
     id:           %r{\d*},
+    session:      %r{[^\/]*},
     login:        %r{[^\/]*},
```
Comment (dgdavid) [inline src/api/config/routes.rb]: Note that I also restored, this constraint for the same reason that `get` login/logout routes. Is it right?

**Comment by dgdavid** [comment]
Thank you @Ana06 for your review, very helpful.

I did the requested changes and I also rewrote commit descriptions to do them better according to your recommendation. But I am not sure if the goal was achieved. If not, let me know and I could retried :)

File: src/api/app/controllers/webui/session_controller.rb [hunk 13]
```diff
@@ -8,36 +8,44 @@ def new; end
   def create
     user = User.find_with_credentials(params[:username], params[:password])
 
-    if user && !user.is_active?
-      redirect_to(root_path, error: 'Your account is disabled. Please contact the administrator for details.')
+    unless user.present?
```
Comment (Ana06) [inline src/api/app/controllers/webui/session_controller.rb]: you don't need the `present?`, user is an object so that won't have any effect. :wink:

File: src/api/config/routes.rb [hunk 14]
```diff
@@ -329,7 +330,6 @@ def self.public_or_about_path?(request)
       get 'user/register_user' => :register_user
 
       get 'user/login' => :login
-      post 'user/logout' => :logout
```
Comment (Ana06) [inline src/api/config/routes.rb]: you say in your commit message

> but keeping the former ones for backward compatibility.

but you are removing them, right? :confused:

File: src/api/app/controllers/webui/session_controller.rb (hunk 13 above)
Comment (Ana06) [reply src/api/app/controllers/webui/session_controller.rb]: Rubocop is failing also because of that, but removing the `present` should make Rubocop happy as well :joy:

File: src/api/config/routes.rb (hunk 14 above)
Comment (dgdavid) [reply src/api/config/routes.rb]: I only restored the `get` routes, as was requested in https://github.com/openSUSE/open-build-service/pull/4731#discussion_r177054886.

Shall I also restore `post` route or reword the commit?

**Comment by dgdavid** [comment]
Thanks again @Ana06 :)

Changes are done.

File: src/api/app/controllers/webui/session_controller.rb (hunk 13 above)
Comment (dgdavid) [reply src/api/app/controllers/webui/session_controller.rb]: Ups! I forgot to run linters yesterday 😅

File: src/api/config/routes.rb (hunk 14 above)
Comment (Ana06) [reply src/api/config/routes.rb]: I think we need to keep all the routes, because other services can be using them. So, restore `post` :wink:

**Comment by dgdavid** [comment]
@Ana06 

Requested changes were done, except the "FIXME" annotation because of it not clear to me who should add it.

**Comment by dgdavid** [comment]
@Ana06, @ChrisBr 


I updated again the first commit after realizing that simply restoring the routes was not enough because those actions are not available anymore in the `UserController`. 

Hence, I think there are two options

- To delegate them to be processed by `SessionController` in a transparent manner.
- To use a redirection in order to advise against using them.

I chose the first option, "mapping" routes to be processed by the appropriate actions of `SessionController`. I also added a few routing tests related to both, former and current login/logout routes.

What do you think?

**Comment by dgdavid** [comment]
@ChrisBr 

> Though I would drop the UserSession model and handle this in the controller. Given that it's only sets / clears Thread.current.

Actually, it was added having in mind to use it as a replacement of `User.current` and `User.current=` methods<sup>[1]</sup>, changes that are not included due to the PR scope. However, I keep it because, IMO, checking and raising possible errors/exceptions there helps to have a more clean controller.

That said, I agree that, ignoring the advantages on the controller side, we could consider it out of the PR scope.

Do you want to remove it?

---
[1] I have a kind of a continue of this branch at https://github.com/dgdavid/open-build-service/tree/feature/remove-current-methods-from-user, where those methods [are removed](https://github.com/dgdavid/open-build-service/commit/81d72ccea947f4e31e42914a17f77dcc5d69cc9d).

**Comment by hennevogel** [comment]
> Do you want to remove it?

Yes please, it's always a good idea to keep the amount of conceptual changes in a PR small...

**Comment by dgdavid** [comment]
Hey @Ana06, @bgeuken 

I would like to know if there is anything more that I could do to finish this PR?

File: src/api/spec/routing/webui/users/requests_spec.rb [hunk 15]
```diff
@@ -1,9 +1,26 @@
...
 
-RSpec.describe '/users/:user/requests routes' do
+RSpec.describe '/users/:user/requests routes', type: :routing do
   let(:user) { Faker::Name.first_name }
+
   it do
     expect(get("/users/#{user}/requests"))
       .to route_to('webui/users/bs_requests#index', user: user)
   end
+
```
Comment (mdeniz) [inline src/api/spec/routing/webui/users/requests_spec.rb]: I don't know why did you added this route specs here in this file. This file is for BsRequest related routes, not for Http request in general...

And also I don't think we need to test the aliases for those routes. It doesn't add too much.

File: src/api/config/routes.rb (hunk 4 above)
Comment (hennevogel) [reply src/api/config/routes.rb]: Because we are looking at the source code all the time. At old issues, not so much.

Comment (hennevogel) [reply src/api/config/routes.rb]: @Ana06 what I don't get is which API you are talking about? This is a WebUI controller. Not an API controller.

File: src/api/config/routes.rb (hunk 12 above)
Comment (bgeuken) [reply src/api/config/routes.rb]: @dgdavid This would only be needed if there would be a parameter named ` session `

**Comment by dgdavid** [comment]
> You have to rebase on master and solve the conflicts at least

Done :)

File: src/api/config/routes.rb (hunk 12 above)
Comment (dgdavid) [reply src/api/config/routes.rb]: Ups! :innocent: Thanks for clarification. I am going to remove it.

File: src/api/config/routes.rb (hunk 4 above)
Comment (mdeniz) [reply src/api/config/routes.rb]: Hey, @dgdavid @hennevogel is right, those are routes for the webui. You can change those without having to respect any backward compatibility.

So, just change it back to what you did in first place. BTW the route tests you added are nonsense right now, but in general we don't test routes :+1:

Comment (dgdavid) [reply src/api/config/routes.rb]: 👍

**Comment by marcus-h** [comment]
On 2018-04-04 02:14:05 -0700, Moisés Déniz Alemán wrote:
> mdeniz commented on this pull request.
> 
> 
> 
> > @@ -328,10 +328,6 @@ def self.public_or_about_path?(request)
>        post 'user/register' => :register
>        get 'user/register_user' => :register_user
>  
> -      get 'user/login' => :login
> 
> Hey, @dgdavid @hennevogel is right, those are routes for the webui. You can change those without having to respect any backward compatibility.
> 
Hmm just a general question (not really related to this PR): why don't
don't we care for backward compatibility for the webui api/routes? For
instance, changing webui routes might break users' bookmarks etc.

**Comment by hennevogel** [comment]
>  why don't don't we care for backward compatibility for the webui api/routes?

We do where it makes sense. It doesn't always make sense.

In this case for instance the moved urls are not pages (well login is but it's not linked from anywhere in the app so nobody really knows about it).

File: src/api/config/routes.rb (hunk 4 above)
Comment (dgdavid) [reply src/api/config/routes.rb]: @mdeniz, 

done!

**Comment by mdeniz** [comment]
Nice done!
Lets merge it after codecov has done it's part

**Comment by Ana06** [comment]
@mdeniz 

> Lets merge it after codecov has done it's part

will that ever happen? :rofl: Codecov is quite "lazy" lately... :unamused:

**Comment by mdeniz** [comment]
Ok, lets merge it... Thanks @dgdavid
